bin/knc.c

```diff
@@ -1072,42 +1111,50 @@ move_data(work_t *work)
 		if (reap() > 0)
 			LOG(LOG_NOTICE, ("child died before EOF"));
 
-		FD_ZERO(&rdset);
-		FD_ZERO(&wrset);
-
-		/* Add stderr */
-		if (work->local_err != -1)
-			FD_SET(work->local_err, &rdset);
+		pfd_net->events = POLLRDHUP;
+		pfd_in->events  = POLLRDHUP;
+		pfd_out->events = 0;
+		pfd_err->events = POLLRDHUP;
 
 		/*
 		 * We have this timeout only to allow us to recover
 		 * from children which prematurely exit
 		 */
-		tv.tv_sec = 30;
-		tv.tv_usec = 0;
+		timeout = POLLRDHUP ? -1 : 30000; /* milliseconds */
```

Comment:
See above comment about defaulting `POLLRDHUP` to `POLLHUP`.

Reply:
Yeah, for now just leave the timeout as `30000`, but then, see above comment about catching `SIGCHLD`: if we do the signal handler self-pipe thing then we can have the `poll(2)` timeout be `0` always.

Reply:
I don't think we need a timeout if we handle SIGCHLD and take action to avoid blocking.  Similar to HTTP, we probably want to read and sink data from the network (up to some limit) to give the client a chance to read any pending response.

Reply:
If we are waiting for events on the sockets, whether or not we spawned a child on a socketpair() should not matter.  I think that the changes made along with the conversion to use poll() obviate the need for the timeout, as long as I also make a change to waitpid() the child in the signal handler.  We might instead choose to SIG_IGN for SIGCHLD, which is allowed in POSIX.1-2008.

Reply:
I pushed an additional commit which waitpid() for SIGCHLD in the signal handler, so that debug logging of the reaped child occurs.